File: Battelle.EPA.WideAreaDecon.Client/src/implementations/parameter/scenario/ContaminatedBuildingType.ts
import { JsonProperty } from 'typescript-json-serializer';
import ParameterType from '@/enums/parameter/parameterTypes';
import DistributionParameterDeserializer from '@/serialization/parameter/DistributionParameterDeserializer';
import IParameter from '@/interfaces/parameter/IParameter';
import DistributionParameter from '../distribution/DistributionParameter';
import NullParameter from '../NullParameter';
import ParameterMetaData from '../ParameterMetaData';

export default class ContaminatedBuildingType implements IParameter {
  @JsonProperty()
  name: string;

  @JsonProperty()
  type: ParameterType = ParameterType.contaminatedBuildingType;

  @JsonProperty(DistributionParameterDeserializer)
  area: DistributionParameter | NullParameter;

  @JsonProperty()
  metaData: ParameterMetaData;

  public isSet(): boolean {
    if (this.area) {
      return this.area.isSet();
    }
    return false;
  }

  constructor(name = 'unknown', metaData = new ParameterMetaData(), area?: DistributionParameter) {
    this.name = name;
    this.area = area !== undefined ? area : new NullParameter();
    this.metaData = metaData;
  }

  isEquivalent(other: IParameter): boolean {
    return this.compareValues(other as ContaminatedBuildingType);
  }

  compareValues(other?: ContaminatedBuildingType): boolean {
    return other ? this.type === other.type && this.area.isEquivalent(other.area) : false;
  }
}

File: Battelle.EPA.WideAreaDecon.API/Models/Parameter/Scenario/ContaminatedBuildingType.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Xml.Linq;
using System.Threading.Tasks;
using System.Runtime.Serialization;
using Battelle.EPA.WideAreaDecon.API.Enumeration.Parameter;
using Battelle.EPA.WideAreaDecon.API.Interfaces.Parameter;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using NPOI.SS.UserModel;

namespace Battelle.EPA.WideAreaDecon.API.Models.Parameter.Scenario
{
    /// <summary>
    /// Implementation of the contaminated building type
    /// </summary>
    public class ContaminatedBuildingType : IParameter
    {
        private static int NameLocation => 2;
        private static int AreaLocation => 6;

        public string Name { get; set; }

        [JsonConverter(typeof(StringEnumConverter))]
        public ParameterType Type => ParameterType.ContaminatedBuildingType;

        public double? Area { get; set; }

        public ParameterMetaData MetaData { get; set; }

        public static ContaminatedBuildingType FromExcel(IRow information)
        {
            ContaminatedBuildingType contaminatedBuilding = new ContaminatedBuildingType();

            double? areaValue = contaminatedBuilding.ParseValueString(AreaLocation, information);

            return new ContaminatedBuildingType()
            {
                Name = information.GetCell(NameLocation)?.ToString() ?? throw new SerializationException("Parameter has no name associated with it in Excel"),
                Area = areaValue,
                MetaData = ParameterMetaData.FromExcel(information)
            };
        }

        public static ContaminatedBuildingType FromEfficacyExcelSheet(IRow information)
        {
            throw new NotImplementedException();
        }

        private double? ParseValueString(int location, IRow information)
        {
            double? value = null;

            var valueString = information.GetCell(location)?.ToString();

            if (!string.IsNullOrWhiteSpace(valueString)) value = double.Parse(valueString);

            return value;
        }
    }
}
